File: app/src/main/java/me/ykrank/s1next/view/fragment/FavouriteListFragment.java
package me.ykrank.s1next.view.fragment;

import android.net.Uri;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;

import com.github.ykrank.androidautodispose.AndroidRxDispose;
import com.github.ykrank.androidlifecycle.event.FragmentEvent;

import javax.inject.Inject;

import me.ykrank.s1next.App;
import me.ykrank.s1next.R;
import me.ykrank.s1next.data.User;
import me.ykrank.s1next.data.api.Api;
import me.ykrank.s1next.data.api.ApiFlatTransformer;
import me.ykrank.s1next.data.api.S1Service;
import me.ykrank.s1next.data.api.UserValidator;
import me.ykrank.s1next.util.IntentUtil;
import me.ykrank.s1next.util.L;
import me.ykrank.s1next.util.RxJavaUtil;
import me.ykrank.s1next.view.event.FavoriteRemoveEvent;
import me.ykrank.s1next.widget.RxBus;

/**
 * A Fragment includes {@link android.support.v4.view.ViewPager}
 * to represent each page of favourite lists.
 */
public final class FavouriteListFragment extends BaseViewPagerFragment {

    public static final String TAG = FavouriteListFragment.class.getName();

    @Inject
    RxBus mRxBus;
    @Inject
    UserValidator mUserValidator;
    @Inject
    User mUser;
    @Inject
    S1Service s1Service;

    private CharSequence mTitle;

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        App.getAppComponent().inject(this);
        super.onViewCreated(view, savedInstanceState);
        L.leaveMsg("FavouriteListFragment");

        mTitle = getText(R.string.favourites);

        mRxBus.get()
                .ofType(FavoriteRemoveEvent.class)
                .to(AndroidRxDispose.withObservable(this, FragmentEvent.DESTROY_VIEW))
                .subscribe(event -> {
                    // reload when favorite remove
                    ApiFlatTransformer.flatMappedWithAuthenticityToken(s1Service, mUserValidator, mUser,
                            token -> s1Service.removeThreadFavorite(token, event.getFavId()))
                            .compose(RxJavaUtil.iOTransformer())
                            .to(AndroidRxDispose.withObservable(this, FragmentEvent.DESTROY_VIEW))
                            .subscribe(wrapper -> {
                                showShortSnackbar(wrapper.getResult().getMessage());
                                loadViewPager();
                            }, this::onError);
                });
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        inflater.inflate(R.menu.fragment_favourites, menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menu_browser:
                IntentUtil.startViewIntentExcludeOurApp(getContext(), Uri.parse(
                        Api.getFavouritesListUrlForBrowser(getCurrentPage() + 1)));

                return true;
            case R.id.menu_favourites_remove:
                showLongSnackbar(R.string.how_to_remove_favourites);
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Override
    BaseFragmentStatePagerAdapter getPagerAdapter(FragmentManager fragmentManager) {
        return new FavouriteListPagerAdapter(fragmentManager);
    }

    @Override
    CharSequence getTitleWithoutPosition() {
        return mTitle;
    }

    /**
     * Returns a Fragment corresponding to one of the pages of favourites.
     */
    private class FavouriteListPagerAdapter extends BaseFragmentStatePagerAdapter {

        private FavouriteListPagerAdapter(FragmentManager fm) {
            super(fm);
        }

        @Override
        public Fragment getItem(int i) {
            return FavouriteListPagerFragment.newInstance(i + 1);
        }
    }
}

File: app/src/main/java/me/ykrank/s1next/view/fragment/FavouriteListFragment.kt
package me.ykrank.s1next.view.fragment

import android.net.Uri
import android.os.Bundle
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import androidx.fragment.app.FragmentManager
import com.github.ykrank.androidautodispose.AndroidRxDispose
import com.github.ykrank.androidlifecycle.event.FragmentEvent
import com.github.ykrank.androidtools.ui.LibBaseViewPagerFragment
import com.github.ykrank.androidtools.util.RxJavaUtil
import com.github.ykrank.androidtools.widget.EventBus
import com.google.android.material.snackbar.Snackbar
import me.ykrank.s1next.App
import me.ykrank.s1next.R
import me.ykrank.s1next.data.api.Api
import me.ykrank.s1next.data.api.ApiFlatTransformer
import me.ykrank.s1next.data.api.S1Service
import me.ykrank.s1next.util.IntentUtil
import me.ykrank.s1next.view.event.FavoriteRemoveEvent
import javax.inject.Inject

/**
 * A Fragment includes [android.support.v4.view.ViewPager]
 * to represent each page of favourite lists.
 */
class FavouriteListFragment : BaseViewPagerFragment() {
    @Inject
    internal lateinit var mEventBus: EventBus
    @Inject
    internal lateinit var s1Service: S1Service

    private var mTitle: CharSequence? = null

    override fun getTitleWithoutPosition(): CharSequence? {
        return mTitle
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        App.appComponent.inject(this)
        super.onViewCreated(view, savedInstanceState)
        leavePageMsg("FavouriteListFragment")

        mTitle = getText(R.string.favourites)

        mEventBus.get()
                .ofType(FavoriteRemoveEvent::class.java)
                .to(AndroidRxDispose.withObservable(this, FragmentEvent.DESTROY_VIEW))
                .subscribe { event ->
                    // reload when favorite remove
                    ApiFlatTransformer.flatMappedWithAuthenticityToken(s1Service, mUserValidator, mUser
                    ) { token -> s1Service.removeThreadFavorite(token, event.favId) }
                            .compose(RxJavaUtil.iOSingleTransformer())
                            .to(AndroidRxDispose.withSingle(this, FragmentEvent.DESTROY_VIEW))
                            .subscribe({ wrapper ->
                                showSnackbar(wrapper.result.message)
                                loadViewPager()
                            }, { this.onError(it) })
                }
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, inflater)
        inflater.inflate(R.menu.fragment_favourites, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.menu_browser -> {
                IntentUtil.startViewIntentExcludeOurApp(requireContext(), Uri.parse(
                        Api.getFavouritesListUrlForBrowser(currentPage + 1)))

                return true
            }
            R.id.menu_favourites_remove -> {
                showSnackbar(R.string.how_to_remove_favourites, Snackbar.LENGTH_LONG)
                return true
            }
            else -> return super.onOptionsItemSelected(item)
        }
    }

    override fun getPagerAdapter(fragmentManager: FragmentManager): FragmentStatePagerAdapter<*> {
        return FavouriteListPagerAdapter(childFragmentManager)
    }

    /**
     * Returns a Fragment corresponding to one of the pages of favourites.
     */
    private inner class FavouriteListPagerAdapter(fm: FragmentManager)
        : LibBaseViewPagerFragment.FragmentStatePagerAdapter<FavouriteListPagerFragment>(fm) {

        override fun getItem(i: Int): FavouriteListPagerFragment {
            return FavouriteListPagerFragment.newInstance(i + 1)
        }
    }

    companion object {

        val TAG = FavouriteListFragment::class.java.simpleName
    }
}
